File: screen.py
import os
import numpy as np

np.set_printoptions(threshold=np.inf)

S = np.full((29, 149)," ")
S[10,:90] = "-"
S[18,:90] = "-"
S[11:17,33:37] = "_"
S[11:17,78:82] = "_"
S[:,90] = "|"
S[:,148] = "|"

#(9, 14)
pump = np.array([
  [' ',' ',' ','.','-','\"','\"','\"','\"','-','.',' ',' ',' '],
  [' ',' ',' ','|','=','=',' ',' ','=','=','|','-','.',' '],
  [' ',' ',' ','|','~','~',' ','~','~','~','|','`','\\','\\'],
  [' ',' ',' ','|','L','I','L','I','L','I','|',' ','|','|'],
  [' ',' ',' ','|',' ',' ',' ',' ',' ',' ','|','/','/',' '],
  [' ',' ',' ','|',' ',' ',' ',' ',' ',' ','|','/',' ',' '],
  [' ',' ',' ','|',' ',' ',' ',' ',' ',' ','|',' ',' ',' '],
  [' ','_','_','|','_','_','_','_','_','_','|','_','_',' '],
  ['[','_','_','_','_','_','_','_','_','_','_','_','_',']']
])

#(5, 21)
car = np.array([
[' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' '],
[' ',' ',' ',' ',' ',' ','.','-','-','-','-','-','-','-','-','.',' ',' ',' ',' ',' '],
[' ','_','_','_','_','/','_','_','_','_','_','|','_','_','_',' ','\\','_','_','_',' '],
['O',' ',' ',' ',' ','_',' ',' ',' ','-',' ','|',' ',' ',' ','_',' ',' ',' ',',','*'],
[' ','\'','-','-','(','_',')','-','-','-','-','-','-','-','(','_',')','-','-','\'',' ']
])

P = np.array([
['#','#','#','#','#','#','#'],
['#','#',' ',' ',' ','#','#'],
['#','#','#','#','#','#','#'],
['#','#',' ',' ',' ',' ',' '],
['#','#',' ',' ',' ',' ',' ']
])

#(9, 27)
parking_pump = np.full((9, 27),"#")
parking_pump[1:-1,1:-1] = " "

#(9, 27)
parking = np.full((9, 27),"+")
parking[1:-1,1:-1] = " "

def printScreen():
  for l in S:
    print("".join(l))

def placeScreen(O, y, x):
  S[y:O.shape[0]+y, x:O.shape[1]+x] = O

cfx1 = 3
cfx2 = 48
cpx1 = 95
cpx2 = 123
cy1 = 2
cy2 = 22
cry = 12

def refresh():
  os.system('clear')
  printScreen()
  S[10,:90] = "-"
  S[18,:90] = "-"
  S[11:17,33:37] = "_"
  S[11:17,78:82] = "_"
  S[:,90] = "|"
  S[:,148] = "|"
  placeScreen(P, 12, 116)

placeScreen(parking_pump, 0, 0)
placeScreen(parking_pump, 20, 0)
placeScreen(parking_pump, 0, 45)
placeScreen(parking_pump, 20, 45)

placeScreen(pump, 0, 28)
placeScreen(pump, 20, 28)
placeScreen(pump, 0, 73)
placeScreen(pump, 20, 73)

placeScreen(parking, 0, 92)
placeScreen(parking, 20, 92)
placeScreen(parking, 0, 120)
placeScreen(parking, 20, 120)
placeScreen(P, 12, 116)

def showLevel(tank):
  C = car.copy()
  level = str(tank) + '%'
  if tank < 10:
    level = '  ' + level
  elif tank < 100:
    level = ' ' + level
  C[0, 9:13] = np.array(list(level)) 
  return C

def ride(x, tank):                                  #3, 48, 95, 123
  placeScreen(showLevel(tank), cry, x)
  refresh()

def emptyRide(x):                                   #3, 48, 95, 123
  placeScreen(np.full(car.shape," "), cry, x)
  refresh()

def park(n, tank):                                  #1 a 4 - bombas, 5 a 8 - vagas
  if n == 1:
    placeScreen(showLevel(tank), cy1, cfx1)
  elif n == 2:
    placeScreen(showLevel(tank), cy2, cfx1)
  elif n == 3:
    placeScreen(showLevel(tank), cy1, cfx2)
  elif n == 4:
    placeScreen(showLevel(tank), cy2, cfx2)
  elif n == 5:
    placeScreen(showLevel(tank), cy1, cpx1)
  elif n == 6:
    placeScreen(showLevel(tank), cy2, cpx1)
  elif n == 7:
    placeScreen(showLevel(tank), cy1, cpx2)
  elif n == 8:
    placeScreen(showLevel(tank), cy2, cpx2)
  refresh()

def empty(n):
  if n == 1:
    placeScreen(np.full(car.shape," "), cy1, cfx1)
  elif n == 2:
    placeScreen(np.full(car.shape," "), cy2, cfx1)
  elif n == 3:
    placeScreen(np.full(car.shape," "), cy1, cfx2)
  elif n == 4:
    placeScreen(np.full(car.shape," "), cy2, cfx2)
  elif n == 5:
    placeScreen(np.full(car.shape," "), cy1, cpx1)
  elif n == 6:
    placeScreen(np.full(car.shape," "), cy2, cpx1)
  elif n == 7:
    placeScreen(np.full(car.shape," "), cy1, cpx2)
  elif n == 8:
    placeScreen(np.full(car.shape," "), cy2, cpx2)
  refresh()

File: screen.c
#include <stdio.h>
#include <stdlib.h>
#include <ncurses.h>
#include <unistd.h>

#define screenX 149
#define screenY 29

char screen[screenY][screenX];

char pump[9][14] = {
    {' ',' ',' ','.','-','\"','\"','\"','\"','-','.',' ',' ',' '},
    {' ',' ',' ','|','=','=',' ',' ','=','=','|','-','.',' '},
    {' ',' ',' ','|','~','~',' ','~','~','~','|','`','\\','\\'},
    {' ',' ',' ','|','L','I','L','I','L','I','|',' ','|','|'},
    {' ',' ',' ','|',' ',' ',' ',' ',' ',' ','|','/','/',' '},
    {' ',' ',' ','|',' ',' ',' ',' ',' ',' ','|','/',' ',' '},
    {' ',' ',' ','|',' ',' ',' ',' ',' ',' ','|',' ',' ',' '},
    {' ','_','_','|','_','_','_','_','_','_','|','_','_',' '},
    {'[','_','_','_','_','_','_','_','_','_','_','_','_',']'}
};

char car[5][21] = {
    {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,0,' ','.','-','-','-','-','-','-','-','-','.',' ',0,0,0,0},
    {' ','_','_','_','_','/','_','_','_','_','_','|','_','_','_',' ','\\','_','_','_',' '},
    {'O',' ',' ',' ',' ','_',' ',' ',' ','-',' ','|',' ',' ',' ','_',' ',' ',' ',',','*'},
    {' ','\'','-','-','(','_',')','-','-','-','-','-','-','-','(','_',')','-','-','\'',' '}
};

char reverseCar[5][21] = {
    {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
    {0,0,0,0,' ','.','-','-','-','-','-','-','-','-','.',' ',0,0,0,0,0},
    {' ','_','_','_','/',' ','_','_','_','|','_','_','_','_','_','\\','_','_','_','_',' '},
    {'*',',',' ',' ',' ','_',' ',' ',' ','|',' ','-',' ',' ',' ','_',' ',' ',' ',' ','O'},
    {' ','\'','-','-','(','_',')','-','-','-','-','-','-','-','(','_',')','-','-','\'',' '}
};

char clearCar[5][21] = {
    {' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' '},
    {' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' '},
    {' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' '},
    {' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' '},
    {' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' '}
};

char P[5][7] = {
    {'#','#','#','#','#','#','#'},
    {'#','#',' ',' ',' ','#','#'},
    {'#','#','#','#','#','#','#'},
    {'#','#',' ',' ',' ',' ',' '},
    {'#','#',' ',' ',' ',' ',' '}
};

char parking_pump[9][27];
char parking[9][27];

void placeScreen(int sizeY, int sizeX, int sY, int sX, char O[sizeY][sizeX]) {
    for (int y = 0; y < sizeY; y++) {
        for (int x = 0; x < sizeX; x++) {
            if (O[y][x] != 0) {
                screen[sY + y][sX + x] = O[y][x];
            }
        }
    }
}

void restock() {
    //inicializa a tela
    for (int x = 0; x < 90; x++) {
        screen[10][x] = '-';
        screen[18][x] = '-';
    }
    for (int y = 11; y < 17; y++) {
        for (int x = 33; x < 37; x++) {
            screen[y][x] = '_';
        }
        for (int x = 78; x < 82; x++) {
            screen[y][x] = '_';
        }
    }
    for (int y = 0; y < screenY; y++) {
        screen[y][90] = '|';
        screen[y][148] = '|';
    }

    //parking
    for (int y = 0; y < 9; y++) {
        for (int x = 0; x < 27; x++) {
            if (x == 0 || x == 26 || y == 0 || y == 8) {
                parking_pump[y][x] = '#';
                parking[y][x] = '+';
            }
        }
    }
    
    //posiciona os objetos
    placeScreen(9, 27, 0, 0, parking_pump);
    placeScreen(9, 27, 20, 0, parking_pump);
    placeScreen(9, 27, 0, 45, parking_pump);
    placeScreen(9, 27, 20, 45, parking_pump);

    placeScreen(9, 14, 0, 28, pump);
    placeScreen(9, 14, 20, 28, pump);
    placeScreen(9, 14, 0, 73, pump);
    placeScreen(9, 14, 20, 73, pump);

    placeScreen(9, 27, 0, 92, parking);
    placeScreen(9, 27, 20, 92, parking);
    placeScreen(9, 27, 0, 120, parking);
    placeScreen(9, 27, 20, 120, parking);
    placeScreen(5, 7, 12, 116, P);
}

void inicialize() {
    restock();    
    initscr();
}

void end() {
    getch();
    endwin();
}

void printScreen() {
    erase();
    for (int y = 0; y < screenY; y++) {
        for (int x = 0; x < screenX; x++) {
            move(y, x);
            printw("%c", screen[y][x]);
        }
    }
    refresh();
    restock();
};

void changeLevel(int tank) {
    int c = tank/100;
    int d = tank/10 - c*10;
    int u = tank - c*100 - d*10;

    car[0][9] = c + '0';
    car[0][10] = d + '0';
    car[0][11] = u + '0';
    car[0][12] = 37;

    reverseCar[0][9] = c + '0';
    reverseCar[0][10] = d + '0';
    reverseCar[0][11] = u + '0';
    reverseCar[0][12] = 37;
}

void ride(int x, int tank, bool reverse) {
    changeLevel(tank);
    if (reverse == true) {
        placeScreen(5, 21, 12, x, reverseCar);
    }
    else {
        placeScreen(5, 21, 12, x, car);
    }
}

void emptyRide(int x) {
    placeScreen(5, 21, 12, x, clearCar);
    restock();
}

void park(int n, int tank) {
    changeLevel(tank);
    switch (n)
    {
    case 0:
        placeScreen(5, 21, 2, 3, car);
        break;
    case 1:
        placeScreen(5, 21, 22, 3, car);
        break;
    case 2:
        placeScreen(5, 21, 2, 48, car);
        break;
    case 3:
        placeScreen(5, 21, 22, 48, car);
        break;
    case 4:
        placeScreen(5, 21, 2, 95, car);
        break;
    case 5:
        placeScreen(5, 21, 22, 95, car);
        break;
    case 6:
        placeScreen(5, 21, 2, 123, car);
        break;
    case 7:
        placeScreen(5, 21, 22, 123, car);
        break;
    default:
        break;
    }
}

void empty(int n) {
    switch (n)
    {
    case 0:
        placeScreen(5, 21, 2, 3, clearCar);
        break;
    case 1:
        placeScreen(5, 21, 22, 3, clearCar);
        break;
    case 2:
        placeScreen(5, 21, 2, 48, clearCar);
        break;
    case 3:
        placeScreen(5, 21, 22, 48, clearCar);
        break;
    case 4:
        placeScreen(5, 21, 2, 95, clearCar);
        break;
    case 5:
        placeScreen(5, 21, 22, 95, clearCar);
        break;
    case 6:
        placeScreen(5, 21, 2, 123, clearCar);
        break;
    case 7:
        placeScreen(5, 21, 22, 123, clearCar);
        break;
    default:
        break;
    }
}
